openwisp_controller/pki/api/views.py

```diff
@@ -0,0 +1,76 @@
+from django.http import HttpResponse
+from django.shortcuts import get_object_or_404
+from django.utils.translation import ugettext_lazy as _
+from django_x509.settings import CRL_PROTECTED
+from rest_framework import pagination
+from rest_framework.authentication import SessionAuthentication
+from rest_framework.generics import ListCreateAPIView, RetrieveUpdateDestroyAPIView
+from rest_framework.permissions import DjangoModelPermissions, IsAuthenticated
+from swapper import load_model
+
+from openwisp_users.api.authentication import BearerAuthentication
+from openwisp_users.api.mixins import FilterByOrganizationManaged
+
+from .serializers import (
+    CaDetailSerializer,
+    CaListSerializer,
+    CertDetailSerializer,
+    CertListSerializer,
+)
+
+Ca = load_model('django_x509', 'Ca')
+Cert = load_model('django_x509', 'Cert')
+
+
+class ListViewPagination(pagination.PageNumberPagination):
+    page_size = 10
+    page_size_query_param = 'page_size'
+    max_page_size = 100
+
+
+class ProtectedAPIMixin(FilterByOrganizationManaged):
+    authentication_classes = [BearerAuthentication, SessionAuthentication]
+    permission_classes = [
+        IsAuthenticated,
+        DjangoModelPermissions,
+    ]
+
+
+class CaListCreateView(ProtectedAPIMixin, ListCreateAPIView):
+    serializer_class = CaListSerializer
+    queryset = Ca.objects.order_by('-created')
+    pagination_class = ListViewPagination
+
+
+class CaDetailView(ProtectedAPIMixin, RetrieveUpdateDestroyAPIView):
+    serializer_class = CaDetailSerializer
+    queryset = Ca.objects.all()
+
+
+def crl_download_view(request, pk):
```

Comment:
Isn't this function doing the same job as `[django-x509 crl_view](https://github.com/openwisp/django-x509/blob/master/django_x509/base/admin.py#L135-L143)`?
Can we please try to avoid duplicating this code? :smile:

Reply:
Yes, I tried to import it, but didn't work. I will try again. :+1:

Reply:
Remember you can change the code in django-x509 to move the required function out of that class and create a mixin! :smile:

Reply:
Hi, @atb00ker I tried creating a mixin in django-x509, and it works well, but when I import it in drf view I get error's regarding arguments.

Reply:
Sorry, I missed this message in mail.
Sure, I can help you out with it. Can you share the stacktrace that you are seeing? `error's regarding arguments` is not enough to guess what the error is ...

Reply:
Meanwhile, I wrote this working code:
```
from django_x509 import settings as app_settings
from django.http import HttpResponse
from django.shortcuts import get_object_or_404
from django.utils.translation import ugettext_lazy as _
from django.views import View
from ..models import Ca


class CrlDownloadMixin:
    def crl_view(self, request, pk):
        authenticated = request.user.is_authenticated
        authenticated = authenticated() if callable(authenticated) else authenticated
        if app_settings.CRL_PROTECTED and not authenticated:
            return HttpResponse(_('Forbidden'), status=403, content_type='text/plain')
        instance = get_object_or_404(self.model, pk=pk)
        return HttpResponse(
            instance.crl, status=200, content_type='application/x-pem-file'
        )


class CrlDownloadMixinView(View, CrlDownloadMixin):
    model = Ca

    def get(self, request, pk):
        return self.crl_view(request, pk)


crlview = CrlDownloadMixinView.as_view()
```

```
class AbstractCaAdmin(BaseAdmin, CrlDownloadMixin):
   ....
```

```
from .base.views import crlview
from django.urls import path

urlpatterns = [path('api/<pk:int>', crlview, name='crl')]
```

Note: this is only a rough implementation using model as is and using random url path, please polish it before using.

Reply:
Thanx a Lot. Done implementing